### PR TITLE
Decoupling: Move all REST API calls to `wp-story-editor`

### PR DESCRIPTION
**Pull request by mohdsayed**
## Context

Move REST API calls to `wp-story-editor` and remove `@wordpress/api-fetch` package dependency from the core story editor.

## To-do

- [x] Create callbacks for all REST API calls and move to wp-story-editor.
- [x] Pass API callbacks back to `wp-story-editor` via config and use in `APIProvider` actions
- [x] Move relevant test cases.
- [x] Move `flattenFormData` to `wp-story-editor`
- ~Handle cases when some API callbacks have not been provided via configuration.~

## Testing Instructions

- [x] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:

1.

## Checklist

- [ ] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [ ] I have verified accessibility to the best of my abilities ([docs](https://github.com/google/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [ ] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/google/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Partially addresses https://github.com/google/web-stories-wp/issues/2918

### CODE REVIEW

File: packages/wp-story-editor/src/api/media.js
Comment (swissspidy) [inline packages/wp-story-editor/src/api/media.js]: Just move this to `wp-story-editor`?

File: packages/wp-story-editor/src/api/test/api.js
Comment (swissspidy) [inline packages/wp-story-editor/src/api/test/api.js:41]: Since you moved this to `wp-story-editor`, I'd just test the `getMedia` callback directly instead of going through the `APIProvider`.

File: tests/phpunit/integration/data/story_post_content_filtered.json
Comment (mohdsayed) [inline tests/phpunit/integration/data/story_post_content_filtered.json]: Not sure how this change got pulled in the merge conflict. Will have to reset.

File: packages/wp-story-editor/src/index.js
Comment (swissspidy) [inline packages/wp-story-editor/src/index.js]: Why was the order of imports changed here? Now the comment is not accurate anymore.

Aside: it might not be true anyway since some other components are imported way earlier now, so to be actually load those styles earlier this import would probably need to be moved to line ~25, before `External dependencies`

File: tests/phpunit/integration/data/story_post_content_filtered.json
Comment (swissspidy) [inline tests/phpunit/integration/data/story_post_content_filtered.json]: You might want to tell your IDE to not accidentally change indentation of JSON files like this :-)

File: packages/story-editor/src/app/api/apiProvider.js
Comment (swissspidy) [inline packages/story-editor/src/app/api/apiProvider.js]: When would that be the case?

File: packages/story-editor/src/app/api/apiProvider.js
Comment (swissspidy) [inline packages/story-editor/src/app/api/apiProvider.js:212]: Alternative:

```js
actions[name] = () => {};
```

WDYT?

File: packages/wp-story-editor/src/api/index.js
Comment (swissspidy) [inline packages/wp-story-editor/src/api/index.js]: Nit: this exposes `statusCheck` callbacks to `story-editor`, even though it's a `wp-story-editor` specific util.

Comment (swissspidy) [reply packages/wp-story-editor/src/api/index.js]: Same for the `metaboxes` export above

File: packages/story-editor/src/app/api/apiProvider.js
Comment (mohdsayed) [reply packages/story-editor/src/app/api/apiProvider.js]: When some api callbacks are not passed from the configuration. One use case is Playground.

File: packages/story-editor/src/app/api/apiProvider.js
Comment (mohdsayed) [reply packages/story-editor/src/app/api/apiProvider.js:212]: It would be easier to check `undefined`, than to check an empty function.

File: packages/wp-story-editor/src/index.js
Comment (mohdsayed) [reply packages/wp-story-editor/src/index.js]: I messed up in resolving a merge conflict. 😬

File: packages/story-editor/src/app/api/apiProvider.js
Comment (swissspidy) [reply packages/story-editor/src/app/api/apiProvider.js:212]: If it's an empty function we wouldn't have to check for it, no? It would simply be a no-op.

But I suppose that only works well for actions and not selectors.

File: packages/story-editor/src/app/api/apiProvider.js
Comment (swissspidy) [reply packages/story-editor/src/app/api/apiProvider.js]: Would the playground pass `apiCallbacks: { uploadMedia: null, ... }` or so in that case or simply not pass that at all? If it doesn't pass anything then it's not in `Object.entries` anyway

File: packages/story-editor/src/app/api/apiProvider.js
Comment (mohdsayed) [reply packages/story-editor/src/app/api/apiProvider.js:212]: We wouldn't have to check for it if we use the empty function however since they are supposed to return a promise, I was thinking of places where we would need to use `.then()` the code would probably look like this.
```js
    const request = getStoryById();
    if ( request ) {
      request.then( ( post ) => {

      } );
    }
```

and places where it's directly returning the API call like 
https://github.com/google/web-stories-wp/blob/b7412eaad26ad4cf43198806aa1d8dcca7c0588d/packages/story-editor/src/app/story/actions/useAutoSave.js#L45-L55

It would still be an inconsistent return and would need handling.

But by looking at the overall required changes, I think it would need fewer changes to use an empty function. Let me try.

File: packages/story-editor/src/app/api/apiProvider.js
Comment (mohdsayed) [reply packages/story-editor/src/app/api/apiProvider.js]: Good catch, the playground can simply not pass at all. Updated code.

File: packages/wp-story-editor/src/api/index.js
Comment (mohdsayed) [reply packages/wp-story-editor/src/api/index.js]: Noted, and added `@todo`, will remove them as I work on moving meta boxes, StatusCheck, and PostLock components.

Comment (mohdsayed) [reply packages/wp-story-editor/src/api/index.js]: Since those still exist in the `story-editor`, they will be required in `APIProivder` for now.

File: packages/story-editor/src/app/api/apiProvider.js
Comment (mohdsayed) [reply packages/story-editor/src/app/api/apiProvider.js:212]: When I look more closely, it's not working very well. For example, here how would I do an early return if I don't want to call these if `getMedia` was not provided.
```js
fetchMediaStart({ pageToken: p }); 
const trackTiming = getTimeTracker('load_media'); 
```

https://github.com/google/web-stories-wp/blob/b7412eaad26ad4cf43198806aa1d8dcca7c0588d/packages/story-editor/src/app/media/local/useContextValueProvider.js#L82-L95

File: packages/wp-story-editor/src/api/media.js
Comment (mohdsayed) [reply packages/wp-story-editor/src/api/media.js]: Absolutely!! 
Moved in https://github.com/google/web-stories-wp/pull/8966/commits/d6d13ce7fcde9312809e3519592d67c2fa0b1f4c

File: packages/story-editor/src/app/api/apiProvider.js
Comment (mohdsayed) [reply packages/story-editor/src/app/api/apiProvider.js:212]: Since the condition is required to be added in a lot of files, I would do it later in a separate PR after this is merged.

File: packages/wp-story-editor/src/api/test/api.js
Comment (mohdsayed) [reply packages/wp-story-editor/src/api/test/api.js:41]: Done in https://github.com/google/web-stories-wp/pull/8966/commits/81899b274c1cd6502ceede50146193c11cbf71b3

File: packages/story-editor/src/app/api/apiProvider.js
Comment (swissspidy) [reply packages/story-editor/src/app/api/apiProvider.js:212]: That's a good point, thanks for exploring & explaining.

Let's keep it as-is then for now 👍

**Comment by spacedmonkey** [comment]
For context, I have created an issue for making a shared package for dashboard and editor for api interface. This as the other benefit of remove repeated code. https://github.com/google/web-stories-wp/issues/8388 . It was doing this work here, does not make sense to move it to it's own package and share between dashboard and editor?

File: packages/wp-story-editor/src/api/getStory.js
Comment (spacedmonkey) [inline packages/wp-story-editor/src/api/getStory.js]: Let's move this file to constant file, it makes easier to share.

File: packages/wp-story-editor/src/api/getStory.js
Comment (spacedmonkey) [inline packages/wp-story-editor/src/api/getStory.js]: See this line will need to added 

https://github.com/google/web-stories-wp/blob/9d18d5f52f1b8a88f170df6be099b0a9d69e40c2/packages/story-editor/src/app/api/apiProvider.js#L58 now that https://github.com/google/web-stories-wp/pull/8946 is merged.

File: packages/wp-story-editor/src/api/pageTemplate.js
Comment (spacedmonkey) [inline packages/wp-story-editor/src/api/pageTemplate.js]:
```suggestion
  const perPage = 100;
  const apiPath = addQueryArgs(customPageTemplates, {
    context: 'edit',
    per_page: perPage,
    page,
    _web_stories_envelope: true,
  });
```

File: packages/wp-story-editor/src/api/getStory.js
Comment (swissspidy) [reply packages/wp-story-editor/src/api/getStory.js]: Share with whom? This is the only place where this constant is needed/used.

File: packages/wp-story-editor/src/api/storyLock.js
Comment (swissspidy) [inline packages/wp-story-editor/src/api/storyLock.js]: Realized that this trailing slash is missing here (even before). Let's add it. We've had some WAF compatibility issues in the past when the trailing slashes were omitted.

```suggestion
  const path = `${stories}${storyId}/lock/`;
```

File: packages/wp-story-editor/src/api/storyLock.js
Comment (swissspidy) [inline packages/wp-story-editor/src/api/storyLock.js]:
```suggestion
  const path = addQueryArgs(`${stories}${storyId}/lock/`, {
```

**Comment by swissspidy** [comment]
> For context, I have created an issue for making a shared package for dashboard and editor for api interface. This as the other benefit of remove repeated code. #8388 . It was doing this work here, does not make sense to move it to it's own package and share between dashboard and editor?

Yes, but it's too early to work on this though. Such a shared package will likely look quite different and contain more things as things evolve. 

Worth noting that this PR is just one step of a long journey. As we walk (see #2918 for roadmap), we will get a better understanding of what such a shared package will look like.

File: packages/wp-story-editor/src/api/pageTemplate.js
Comment (mohdsayed) [reply packages/wp-story-editor/src/api/pageTemplate.js]: Done in https://github.com/google/web-stories-wp/pull/8966/commits/a332ebafd61dfc4a7f9ceddb84412e739ce7b117

File: packages/wp-story-editor/src/api/storyLock.js
Comment (mohdsayed) [reply packages/wp-story-editor/src/api/storyLock.js]: Done in https://github.com/google/web-stories-wp/pull/8966/commits/a332ebafd61dfc4a7f9ceddb84412e739ce7b117

File: packages/wp-story-editor/src/api/storyLock.js
Comment (mohdsayed) [reply packages/wp-story-editor/src/api/storyLock.js]: Done in https://github.com/google/web-stories-wp/pull/8966/commits/a332ebafd61dfc4a7f9ceddb84412e739ce7b117

File: packages/wp-story-editor/src/api/getStory.js
Comment (mohdsayed) [reply packages/wp-story-editor/src/api/getStory.js]: Done, thanks

**Review by swissspidy — APPROVED** [review]

File: packages/wp-story-editor/src/api/getStory.js
Comment (spacedmonkey) [reply packages/wp-story-editor/src/api/getStory.js]: We have two other post type that follow the same structure, of story data etc. Might be useful in these other places.

File: packages/wp-story-editor/src/api/pageTemplate.js
Comment (spacedmonkey) [inline packages/wp-story-editor/src/api/pageTemplate.js]:
```suggestion
  const path = addQueryArgs(customPageTemplates, {
    context: 'edit',
    per_page: perPage,
    page,
    _web_stories_envelope: true,
  });
  return apiFetch({ path }).then(({ headers, body }) => {
```

File: packages/wp-story-editor/src/api/statusCheck.js
Comment (spacedmonkey) [inline packages/wp-story-editor/src/api/statusCheck.js]: This `base64Encode` function should be moved to this package. It is only used in the API provider and is linked to the WordPress implementation.

File: packages/wp-story-editor/src/api/story.js
Comment (spacedmonkey) [inline packages/wp-story-editor/src/api/story.js]:
```suggestion
export function getStoryById(storyId, stories, isDemo = false) {
  const path = addQueryArgs(`${stories}${storyId}/`, {
    context: 'edit',
    _embed: STORY_EMBED,
    web_stories_demo: isDemo,
    _fields: STORY_FIELDS,
  });

  return apiFetch({ path });
}

export function getDemoStoryById(storyId, stories) {
  return getStoryById(storyId, stories, true);
}
```

Is this a little cleaner?

File: packages/story-editor/src/app/api/apiProvider.js
Comment (spacedmonkey) [inline packages/story-editor/src/app/api/apiProvider.js]: This function is very heavily linked to the WordPress implementation, like `base64Encode` and all the field mapping, why has this function remained here. This function feels like it should be moved.

File: packages/wp-story-editor/src/api/story.js
Comment (swissspidy) [reply packages/wp-story-editor/src/api/story.js]: That would expose the `isDemo` param to `story-editor`

File: packages/wp-story-editor/src/api/getStory.js
Comment (mohdsayed) [reply packages/wp-story-editor/src/api/getStory.js]: Done in https://github.com/google/web-stories-wp/pull/8966/commits/aa896b0221e33b7e4a96f5685a467aae8904e442

File: packages/wp-story-editor/src/api/pageTemplate.js
Comment (mohdsayed) [reply packages/wp-story-editor/src/api/pageTemplate.js]: Done in https://github.com/google/web-stories-wp/pull/8966/commits/951558e5067077554e973489c0745cfbff5ca482

File: packages/wp-story-editor/src/api/statusCheck.js
Comment (mohdsayed) [reply packages/wp-story-editor/src/api/statusCheck.js]: Done in https://github.com/google/web-stories-wp/pull/8966/commits/57e6c7e5626e2065dee39537c42f60881807558f

File: packages/story-editor/src/app/api/apiProvider.js
Comment (mohdsayed) [reply packages/story-editor/src/app/api/apiProvider.js]: Done in https://github.com/google/web-stories-wp/pull/8966/commits/57e6c7e5626e2065dee39537c42f60881807558f

File: packages/wp-story-editor/src/api/story.js
Comment (mohdsayed) [reply packages/wp-story-editor/src/api/story.js]: Done in https://github.com/google/web-stories-wp/pull/8966/commits/25cfc094e9734e4a3f558c0b972dd3ff58c53798

**Comment by mohdsayed** [comment]
@spacedmonkey Just a friendly ping that my other PR/tasks are blocked by this PR to some extent 🙂

**Review by spacedmonkey — APPROVED** [review]

**Pull request merged**
